### DISCUSSION TITLE
crontriggeringpolicy vs timebasedtriggeringpolicy for daily rotation of log files

### DISCUSSION THREAD

**Discussion by rswathirao** [apache/logging-log4j2#2826] [Q&A] [answered]
timebasedtriggeringpolicy we are seeing some issues with file rotation in some scenarios where too many threads on the host are active and though single application is writing to the log:

however not seeing rotation happen in some scenarios

so we wnated to use corntriggeringpolicy 

Internally how is crontriggeringpolicy evalueated, is it for every log entry time is compared against cron expression or just one time for lets say a file tthat need to be rotated at midnight

Anyone has evaluated this. appreciate your insight in advance,,

**Comment by rgoers** [comment, marked as answer]
CronTriggeringPolicy uses the ConfigurationScheduler and schedules an activity at the rollover time. When that occurs it just calls the manager.rollover() method.  TimeBasedTriggeringPolicy compares the time in every log event and rolls over when it gets a time greater than the rollover time.

**Comment by rswathirao** [comment]
@rgoers TY for quick response
so.. the rollover time is evaluated only once for the schedule expression , if log4j config file changed will this be reevaluated each time ? or for every log entry.. for CronTriggeringPolicy 

one other issue i faced during conversion is unable to provide pattern expression as a suffix to active log file
we had tht in log4j1 but not in 2.x?


example in below config:

        <RollingFile name="testLog"
                     fileName="/var/log/TestLog.log"
                     filePattern="/var/log/TestLog.%d{yyyy-MM-dd}.log.gz" append="true">
            <PatternLayout pattern="%d{dd MMM HH:mm:ss.SSS},%m%n"/>
            <Policies>
                 <CronTriggeringPolicy schedule="0 0 0 * * ?" evaluateOnStartup="true"/>
            </Policies>
            <DefaultRolloverStrategy max="10"/>
        </RollingFile>


is there a way to provide pattern for acrive log file name : i.e. fileName to have a pattern like TestLog-%d{yyyy-MM-dd}.log rather then just TestLog.log

**Reply by ppkarwasz** [reply]
> so.. the rollover time is evaluated only once for the schedule expression , if log4j config file changed will this be reevaluated each time ? or for every log entry.. for CronTriggeringPolicy

All the rolling file appenders with the same `fileName` value share a single `RollingFileManager`. This means that all changes to the triggering policies and rollover strategies that you perform after your application started, will be **ignored**. If you want to change triggering policy, you need to restart you application (or change `fileName`).

The purpose of this feature is to prevent multiple appenders (for example the RFA from the old configuration and the RFA from the new configuration) from opening the same file at the same time.

See [rolling file appenders](https://logging.staged.apache.org/log4j/2.x/manual/appenders/rolling-file.html#appenders) for more details.

> is there a way to provide pattern for acrive log file name : i.e. fileName to have a pattern like TestLog-%d{yyyy-MM-dd}.log rather then just TestLog.log

If you want to use the same pattern for the current and archives log files, just remove the `fileName` attribute.
Under the hood the rollover strategy will change from `DefaultRolloverStrategy` to `DirectWriteRolloverStrategy`:

```xml
    <RollingFile name="testLog"
                 filePattern="/var/log/TestLog.%d{yyyy-MM-dd}.log.gz" append="true">
        <PatternLayout pattern="%d{dd MMM HH:mm:ss.SSS},%m%n"/>
        <Policies>
             <CronTriggeringPolicy schedule="0 0 0 * * ?" evaluateOnStartup="true"/>
        </Policies>
        <DirectWriteRolloverStrategy maxFiles="10"/>
    </RollingFile>
```
